Cache expression type when checking assertion

gabritto opened the pull request:
Experiment to see if we can revert the perf impact of #53261.

gabritto commented:
@typescript-bot perf test this faster

gabritto commented:
@typescript-bot perf test this faster

gabritto commented:
@typescript-bot perf test this

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -34547,9 +34548,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+        const { type } = getAssertionTypeAndExpression(node);
         const errNode = isParenthesizedExpression(node) ? type : node;
-        const exprType = getRegularTypeOfObjectLiteral(getBaseTypeOfLiteralType(checkExpression(expression)));
+        const exprType = getRegularTypeOfObjectLiteral(getBaseTypeOfLiteralType(getNodeLinks(node).assertionExpressionType!));
```
Can you hoist the access of `assertionExpressionType` to an `assertDefined` for easier debugging if something goes wrong?

gabritto commented:
@typescript-bot perf test this faster

DanielRosenwasser commented:
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt

DanielRosenwasser commented:
Looks like there's going to be a memory bump for Angular, but I think that is a worthwhile tradeoff.

gabritto commented:
I was doing the comparison of the results of this PR (https://github.com/microsoft/TypeScript/pull/54224#issuecomment-1546322912) with the ones in the original PR that introduces the regression (https://github.com/microsoft/TypeScript/pull/53261#issuecomment-1480195791). It's tricky to compare those directly, but it seems like this PR takes care of most of the regression on `compiler-unions`, but it doesn't seem to improve the perf of `xstate`.

DanielRosenwasser commented:
> I was doing the comparison of the results of this PR (https://github.com/microsoft/TypeScript/pull/54224#issuecomment-1546322912) with the ones in the original PR that introduces the regression (https://github.com/microsoft/TypeScript/pull/53261#issuecomment-1480195791). It's tricky to compare those directly, but it seems like this PR takes care of most of the regression on `compiler-unions`, but it doesn't seem to improve the perf of `xstate`.

I think that's fine - but I have some ideas we can try.

Two optimizations that would be interesting to try in follow-up PRs would me:

1. Get rid of the intermediate helper that gets both the type and expression. It's only used in two places, and unfortunately we can't always depend on the runtime to inline the function and eliminate the allocation.
2. Don't defer the work if either type has any of `TypeFlags.{Any,Unknown,Never}` since those types will always be castable to-and-from each other.

weswigham commented on src/compiler/checker.ts:
```diff
@@ -34516,13 +34516,14 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+            return getRegularTypeOfLiteralType(checkExpression(expression, checkMode));
         }
+        const links = getNodeLinks(node);
+        links.assertionExpressionType = checkExpression(expression, checkMode);
```
By moving the `checkExpression` for the `as const` case into the `if`, it's no longer used in the deferred-check case at all; so you shouldn't *actually* need to cache it like this (which is a tad unsafe, so if it's avoidable, I'd like to avoid it), since you'll either do it once early or once deferred.

DanielRosenwasser replied:
So what're you thinking @weswigham? Place this in an explicit `else`?

DanielRosenwasser replied:
Or I guess, keep it in the same as before, but then cache it on links unconditionally?

That makes sense, you're calling it twice in some cases anyway now, right?

gabritto replied:
If I understand correctly, that's what I was doing in https://github.com/microsoft/TypeScript/pull/53261/files/45124b1447f2ff88bdf2f9e65f0e1766a80a70a3, and that's what caused `xstate` checking perf to increase by 10-13% (https://github.com/microsoft/TypeScript/pull/53261#issuecomment-1472613782).

DanielRosenwasser replied:
@gabritto but aren't you calling `checkExpression` twice now?

DanielRosenwasser replied:
I guess there's no "checking twice" but there's some code deduplication - `checkExpression` is going to happen anyway. I'd rather we got rid of that and go back to having `exprType`

That said, I don't see why avoiding deferral is a problem in that case.

weswigham replied:
>  I'd rather we got rid of that and go back to having exprType

I wouldn't. The `CheckMode`-independent unconditional cache is unsafe and could lead to weird inconsistencies between signature checking states, which can lead to odd editor/command line error inconsistencies. We do it in one or two other places for perf reasons, but it's to be avoided where possible.

weswigham commented:
> Get rid of the intermediate helper that gets both the type and expression. It's only used in two places, and unfortunately we can't always depend on the runtime to inline the function and eliminate the allocation.

We can do this, but the impact should be a minor memory pressure improvement - perf-wise it should be neutral, unless it just barely puts us over the GC line in a project. I'll be surprised if this does too much. If it does.... I'd just call the original change unlucky(?), since almost any new temporary object could have triggered the increased cost.

> Don't defer the work if either type has any of TypeFlags.{Any,Unknown,Never} since those types will always be castable to-and-from each other.

That would require actually checking the expression early, which we can entirely avoid in the non-as-const case which is what we _needed_ to avoid to prevent the circularity error in the first place. For the expression type, anyway. For the assertion type, sure; but, again, just deferring a node's check is pretty free (all it means is we add an id to a set and come back to it later) - in fact, *usually* it's perf-positive to defer a node check because it prevents us from repeating the work when we check an expression tree multiple times during signature resolution (by only doing it once at the end once the signature types are locked in). Avoiding the deferral isn't something I think'll actually be too useful in terms of performance - mostly just complicated. Who knows, I could be wrong, but my hopes aren't high.

Back on this fix, though - If `xstate` still has worse perf even with us only calling `checkExpression` once (during either eager or deferred checking), while `compiler-unions` is fixed, then I'd wager it's, for some reason, impacted in an outsized way by something else that's duplicated between the two stages. The only other work duplicated it looking up the expression and type nodes - if it was JS, I'd buy it could be that (we don't cache jsdoc type node lookups, which may get convoluted), but it's not - it's a TS project. The other thing duplicated is the `getTypeFromTypeNode` call on the asserted type; _usually_ type nodes which have complex types cache their type internally, but some type node kinds don't. Next thing I'd try is caching the `getTypeFromTypeNode` call, so we only resolve the type node into actual types once. _If_ that _is_ the issue, it implies we need better caching within `getTypeFromTypeNode` itself - it implies a broader issue that this just happened to touch, which we'd want to investigate after triaging this (aka: we found a problem that we may be able to fix more broadly for broader perf wins!).

weswigham commented:
It's also possible that the original fix actually just removed circularity errors in `xstate`. The perf suite doesn't report error metrics, but if we fixed some circularities in `xstate`, we'd _expect_ the perf to be worse, since we fixed some bogus `any` types and started doing some real compatibility checking.

DanielRosenwasser commented:
> That would require actually checking the expression early

If I'm not mistaken, I think the expression already is checked early, and unconditionally. The deferred work is checking if the two types are comparable in either direction.

weswigham commented:
> If I'm not mistaken, I think the expression already is checked early, and unconditionally. The deferred work is checking if the two types are comparable in either direction.

You're right, it is. I'm just thinking we can actually defer the expression check now, too in the non-as-const case, which should be a savings.

gabritto commented:
> It's also possible that the original fix actually just removed circularity errors in `xstate`. The perf suite doesn't report error metrics, but if we fixed some circularities in `xstate`, we'd _expect_ the perf to be worse, since we fixed some bogus `any` types and started doing some real compatibility checking.

That sounds like a reasonable explanation for the weird `xstate` behavior, so I'm gonna look into it, since if that's the case, then we can just check the expression in the deferred check and no need to check.

weswigham commented:
> That sounds like a reasonable explanation for the weird xstate behavior, so I'm gonna look into it, since if that's the case, then we can just check the expression in the deferred check and no need to cache types.

I just checked - the `xstate` perf test has 7 errors in 5.0, but only 4 in 5.1.

gabritto commented:
@typescript-bot perf test this faster

gabritto commented:
Ok, I compiled `xstate` with the current main and with a version of this PR where we only call `checkExpression` in the deferred check, and the errors look exactly the same, so it doesn't immediately seem to be the case that calling `checkExpression` only in the deferred check eliminates some earlier error that cause us to do more work, so the mystery remains. 🙁

weswigham commented on src/compiler/checker.ts:
```diff
@@ -34515,31 +34515,35 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+        checkSourceElement(typeNode);
+        const type = getTypeFromTypeNode(typeNode);
+        // See if we need to check whether the expression and asserted types are comparable.
+        if (!isErrorType(type) && !((exprType.flags | type.flags) & (TypeFlags.AnyOrUnknown | TypeFlags.Never))) {
```
IMO, this is silly - this is quite literally what the first lines of `isTypeAssignableTo` does, and all doing it early avoids is putting a node id into a set (in exchange for duplicating this comparison in the cases where that _isn't_ avoided, which is far and away more common).

gabritto commented:
Yep, like Wesley pointed out, latest changes didn't really help. I'll revert.

DanielRosenwasser commented:
Seems reasonable! I wanted to have that in a separate PR anyway.

weswigham commented on src/compiler/checker.ts:
```diff
@@ -34523,6 +34523,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             }
             return getRegularTypeOfLiteralType(exprType);
         }
+        const links = getNodeLinks(node);
```
```suggestion
function checkAssertionWorker(node: JSDocTypeAssertion | AssertionExpression, checkMode: CheckMode | undefined) {
        if (isConstTypeReference(type)) {
            if (!isValidConstAssertionArgument(expression)) {
                error(expression, Diagnostics.A_const_assertions_can_only_be_applied_to_references_to_enum_members_or_string_number_boolean_array_or_object_literals);
            }
            return getRegularTypeOfLiteralType(checkExpression(expression, checkMode));
        }
        checkSourceElement(type);
        checkNodeDeferred(node);
        return getTypeFromTypeNode(type);
 }
```
and no changes in `checkAssertionDeferred` from current `main` should result in the same number of `checkExpression` calls as the unsafe cache here, but without needing an unsafe cache, is what I'm trying to say. :P

DanielRosenwasser replied:
So getting rid of `assertionExpressionType` on `NodeLinks` entirely, just only doing `checkExpression` in `const` contexts and in the deferred check. Makes sense.
